### ISSUE TITLE
CLI threat 

### ISSUE THREAD

**Issue by sdgathman** [turistu/odds-n-ends#1]
Thank you for this detailed report.  This is a real concern for me - I am a heavy CLI user, having used ascii terminal from the 1970s until finally donating my last wyse terminal to a museum this millenium.  (A linux laptop at this point is cheaper, more power efficient, and depending on your terminal app, more secure.)

Why does firefox not consider this a security threat?  Because it is mainly an attack on CLI.  Maybe they only use IDEs for development.  There were similar issues back in the ascii terminal days where a malicious app would send an escape that would load arbitrary text into some terminal [mis]feature to be triggered later.  (Back then it was mostly childish vandalism like `rm -rf *`.)

Is there any way to provide a barrier to clear paste buffers when switching to selected apps (e.g. terminals)?  Or GUIs are supposed to be visual - a warning light or notifier that comes on when javascript writes to the paste buffer?  (Javascript being a primary remote threat of this nature atm - of course any local app can write to a paste buffer.)  

Now that I think of it, I think I saw a paste buffer visualization gui app a while back - can't remember what it was called, but that might address this issue.

**Comment by turistu** [comment]
> Is there any way to provide a barrier to clear paste buffers when switching to selected apps (e.g. terminals)?

Short answer: no

First, there are no barriers between X11 app/clients on the same display. All the X11 clients are supposed to be trusted. Most X11 operations are using *windows* as endpoints on which they set properties and listen for events, but any X11 app can set properties and listen for events on windows created by another client just as if they were its own.

Second, there are no "paste buffers" on X11 [1]. When an app `A` copies something and another app `B` pastes it, `A` transfers the data directly to `B` just when `B` asks for it, it does not write it to some buffer from where any other app can pick it up later. This description glosses over a lot of details (you can go and read chapter 2 of [**icccm.txt**](https://www.x.org/releases/X11R7.7/doc/xorg-docs/icccm/icccm.txt) for the exact mechanism), but the gist of it is that:
1. another app `C` can interfere with the transfer between `A` and `B` at any point and any way it likes
2. yet another app `D` can become the owner of the same selection (and even transfer a different text to another app `E`) while the transfer between `A` and `B` is still in progress.

tl;dr; it's firefox which has to be fixed (or just chucked!), there's no way around it ;-)

[1] there actually still are the so-called "cut buffers" (which are simple properties of the root window), but they're obsoleted since 30 years and they're AFAIK only used as a fallback by xterm.

**Comment by turistu** [comment]
> > Why does firefox not consider this a security threat?
> 
> I too would like to know this.

I'd like to know it too, but they didn't bother to explain it to me either. The most I got from them was a rationale which involved things like "opening a console", "hitting copy" -- i.e. it was obvious that they didn't even run an X11 system or bothered to test it on it. Just like if someone had typed "why is paste jacking not considered a serious issue" in chatGPT ;-)

> Can you perhaps link us to the bug report?

It's https://bugzilla.mozilla.org/show_bug.cgi?id=1855345, but it's still "secret".

In any case, when and if they open it, there were **no other messages sent to them from my part** than those included in that discussion.

> Also, I'm unable to successfully run the demo on Librewolf even with NoScript disabled. It may not be vulnerable?

Maybe, I'll look into it.

> If so, then it's possible a config option may have been changed in Librewolf that prevents it, which could be applied to Firefox?

I don't think there's any option in the stock firefox.

**Comment by quantumpacket** [comment]
I stand corrected. Librewolf is indeed vulnerable :cry:. Looking at the clipboard contents I see the payload is copied. Silly me was not using middle-click, which appears to be a requirement. Good thing for me I never use middle-click :relieved:.

**Comment by sdgathman** [comment]
> First, there are no barriers between X11 app/clients on the same display. All the X11 clients are supposed to be trusted. Most X11 operations are using _windows_ as endpoints on which they set properties and listen for events, but any X11 app can set properties and listen for events on windows created by another client just as if they were its own.

This being the case, an app to visualize the paste buffer would help greatly.  There might already be such a thing, I've just never really wanted it.  I start looking for one.

**Comment by turistu** [comment]
> > First, there are no barriers between X11 app/clients on the same display. All the X11 clients are supposed to be trusted. Most X11 operations are using _windows_ as endpoints on which they set properties and listen for events, but any X11 app can set properties and listen for events on windows created by another client just as if they were its own.
> 
> This being the case, an app to visualize the paste buffer would help greatly. There might already be such a thing, I've just never really wanted it. I start looking for one.

I have written such a thing, but it's a mess and it will take me a while to clean it up. In any case, the basic idea is not complicated: it monitors the selection ownership via the [Xfixes](https://www.x.org/releases/X11R7.7/doc/fixesproto/fixesproto.txt) extension, and then fetches it and writes to stdout in the same manner as `xsel` or `xclip`.

**Comment by quantumpacket** [comment]
> This being the case, an app to visualize the paste buffer would help greatly.

@sdgathman I use Clipman on Xfce, and it shows me what is currently copied, and what has been copied in the past. This is what I see after visiting the PoC.

![Screenshot_2023-10-17_14-50-26](https://github.com/turistu/odds-n-ends/assets/8516784/a00b72cc-3e24-4077-8c44-761f601c6a37)

Not ideal, as you would need to check it everytime you're pasting something.

**Comment by sdgathman** [comment]
I often run firefox and other X apps through X2goclient (not for security, just to do gui things on several boxes from a single console).  This does provide a barrier (because the remote app talks to a different X server), although I haven't mapped out the exact boundaries.  E.g. x2goclient itself is a local X app, so maybe you could still paste into it.

**Comment by aksel** [comment]
If you can live without middle-click paste from Firefox, you can disable it by setting `clipboard.autocopy` to false in `about:config`.

**Comment by turistu** [comment]
> If you can live without middle-click paste from Firefox, you can disable it by setting `clipboard.autocopy` to false in `about:config`.

Just giving up on firefox would be easier ;-) That's like asking someone who has driven automatic all his life to just start using the stick and clutch.

**Comment by aksel** [comment]
> Just giving up on firefox would be easier ;-) That's like asking someone who has driven automatic all his life to just start using the stick and clutch.

lol okay

I'd argue it's more like realizing there's a vulnerability in the remote lock button. Because of this, someone sneaky could theoretically clone your key without you noticing - provided you willingly give them your key and close your eyes real quick :)

I'd just disable the remote lock. It'll be a pain in the ass, and I'll be yelling at the car manufacturer until they take it seriously and fix it. I've also lost a little bit of respect for them, because they haven't been taking it seriously so far. But I like my car, jeez.

And get this - I figured out a quick and easy way to disable the remote lock, and posted about it. And this guy was all like "dude just buy a new car". Can you believe that?? Get real haha. I bet he drives an automatic, too, like a lil dweeb.

**Comment by turistu** [comment]
> I'd argue it's more like realizing there's a vulnerability in the remote lock button. Because of this, someone sneaky could theoretically clone your key without you noticing - provided you willingly give them your key and close your eyes real quick :)

Lame.

> And get this - I figured out a quick and easy way to disable the remote lock, and posted about it.

Seriously, for anybody else reading this, I think that the workaround from my report] (= disable the javascript Selection API from a [content script][1]), is much less intrusive than just totally crippling firefox (and it has the potential of being tailored more narrowly and even made completely transparent, should they not fix the issue any time soon).

[1]: https://addons.mozilla.org/en-US/firefox/addon/disable-the-selection-api/

Turning off `clipboard.autocopy` is only a scare-quote "solution" for people who are **not** using the primary selection at all, who are **not** affected by this issue in any way, yet are somehow still fearful about unknown threats.

> And this guy was all like "dude just buy a new car". Can you believe that?? Get real haha. I bet he drives an automatic, too, like a lil dweeb.

**Comment by aksel** [comment]
Yeah okay, sorry for the snark. But snark begets snark.

In any case, I agree that this is a severe vulnerability in Firefox, one that should be mitigated there. Furthermore, I don't think pastejacking via selection clipboard is the sole exploit.

Say a user selects a command, intending to copy and paste it. What if you then replace what is selected with an invisible malicious command, while rendering a fake selection highlight over the original command? If the user presses CTRL-C, it would end up in the clipboard, right?

In any case, it's disappointing to hear Mozilla do not consider it a vulnerability. It's frankly absolutely bizarre that JavaScript is able to silently and sneakily change what you have selected. Invisible things, too. How can invisible text even be selected?

I use select paste it all the time. But I'll disable it from Firefox for now, because I think switching browsers due to a bug is a tad extreme. Now, if Mozilla continues to dismiss it, then I'd uninstall. And patch it myself and then reinstall lol

**Comment by quantumpacket** [comment]
> Say a user selects a command, intending to copy and paste it. What if you then replace what is selected with an invisible malicious command, while rendering a fake selection highlight over the original command? If the user presses CTRL-C, it would end up in the clipboard, right?

Someone on the OSS-Security mailing list gave another type of possible attack, where say you are copying a crypto address be it Bitcoin or whatever to send funds to and an open tab overwrites it with someone else's address, and now your money is transferred to the wrong wallet. I would say that would impact a lot more people than inserting commands to be pasted into a terminal. This is definitely something Firefox should fix.

**Comment by fi5er** [comment]
Just as a heads up, the OSS-Security mail has recently been [posted to HN](https://news.ycombinator.com/item?id=37964835). There are some interesting observations in the comments. You can probably expect some activity here.

It's a shame that people cannot comment on Bugzilla issues that are tagged as security relevant, but are dismissed by the developers. How are we supposed to let them know we disagree?

**Comment by sdgathman** [comment]
> I'd argue it's more like realizing there's a vulnerability in the remote lock button. Because of this, someone sneaky could theoretically clone your key without you noticing - provided you willingly give them your key and close your eyes real quick :)

There is a real vulnerability with remote lock buttons used by many car manufacturers.  They use a 16bit initialization vector, and a dedicated crook (and yes, there are surprising many) "just" needs to record the lock and unlock keypresses for about 16K cars of the same remote lock brand.   After that dictionary is built, they can record an owner locking their car, send the unlock code while the owner is shopping, and take whatever they want from the car. 

> I'd just disable the remote lock. It'll be a pain in the ass, and I'll be yelling at the car manufacturer until they take it seriously and fix it. I've also lost a little bit of respect for them, because they haven't been taking it seriously so far. But I like my car, jeez.

I just got the cheaper key without remote lock.  Don't need remote lock, there is a button on the inside door handle.
